Affine registration

jboulanger opened the pull request:
## Description

Intensity based affine registration of image pairs.
- estimate a transform to register the moving image using ndi.affine_transform
- supports 2d and 3d images
- supports color image using channel_axis parameter (more flexible for 3d images)
- supports affine, euclidean and translation motion models using a string parameter
- supports weights/mask to use only a part of the image. Modifies the normalized_mutual_information to support weights.
- includes 2 solvers: lucas kanade [1] and studholme [2,3] parametrized as functions callback
- related to PR [3544](https://github.com/scikit-image/scikit-image/issues/3544) and  [7050](https://github.com/scikit-image/scikit-image/issues/7050). The studholme solver is inspired heavily from PR 3544.
- include an quantification of alignment error (target registration error)

The API tries to take inspiration from the PR3455, however I found it not very user friendly to pass callback for handling euclidean or translation motion model. Moreover the logic is different between the two solvers and could not find a way to conciliate the previously discussed API with the added solver. PR3455 was relying on transforms in the example to handle restricted model (euclidean) which is not compatible with ndi.affine_transform,  here we stick to ndi.affine_transform to ba able to handle 3D images.

Comparison bewteen the solvers and cv2.findTransformECC. The 2nd row display the registration error map.
![image](https://github.com/scikit-image/scikit-image/assets/3415561/a99c3bfa-6be5-4c36-93c9-3e4f96ac900b)

Registration of a [brain MRI volume](https://graphics.stanford.edu/data/voldata/mrbrain-8bit.tar.gz)
![image](https://github.com/scikit-image/scikit-image/assets/3415561/e127c85f-8432-4c7e-97a5-2ac4867117fb)

### TODO
- [ ]  euclidean is not implemented for lucas kanade solver for dimension > 2.

### References
1. [algo_affine](http://robots.stanford.edu/cs223b04/algo_affine_tracking.pdf)
2. Studholme C, Hill DL, Hawkes DJ. Automated 3-D registration of MR  and CT images of the head. Med Image Anal. 1996 Jun;1(2):163-75.
3. J. Nunez-Iglesias, S. van der Walt, and H. Dashnow, Elegant SciPy: The Art of Scientific Python. O’Reilly Media, Inc., 2017.




  <!--




- Reference relevant issues or related pull requests with their URL / #<number>.
- Do not use AI to help write your contribution.
- Use `pre-commit` to check and format code.
-->

## Checklist

<!-- Before pull requests can be merged, they should provide: -->

- [x]  A descriptive but concise pull request title
- [x]  [Docstrings for all functions](https://github.com/numpy/numpydoc/blob/main/doc/example.py)
- [x] [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- [x] A gallery example in `./doc/examples` for new features
- [x] [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed (as much as possible)

## Release note

For maintainers and optionally contributors, please refer to the [instructions](https://scikit-image.org/docs/stable/development/contribute.html#documenting-changes) on how to document this PR for the release notes.

```release-note
...
```

Commit:
basic registration function added

leGIT-bot committed:
increased accuracy

added exp input as measure of thoroughness
got rid of off by one errors
added previous closed commit

seanbudd committed:
Refactor and expose module to skimage

seanbudd committed:
fix imports

Commit:
updated register functionality to be similar scipy ch7

Commit:
do callback on iteration of registration

leGIT-bot committed:
allowed different image dimentions

leGIT-bot committed:
Added documentation

leGIT-bot committed:
fixed some nuances

leGIT-bot committed:
some more fine-tuning

seanbudd committed:
various fixes so test runs smoothly

seanbudd committed:
ran autopep8

seanbudd committed:
export p_to_matrix functions as they are needed in the plotting, rename parameters

seanbudd committed:
remove p and matrix stuff

seanbudd committed:
naming fixes

seanbudd committed:
change cost mse

seanbudd committed:
tidied up levels

Commit:
changed to affine transformation

seanbudd committed:
remove basinhopping

seanbudd committed:
fix doc strings

seanbudd committed:
limit levels for gaussian pyramid

Commit:
basic example done

Commit:
code review fixes 1/2

Commit:
added docstring to registration.py

Commit:
added array size

Commit:
minor fixes to registration py

Commit:
make it ndimensional

Commit:
Added unit tests

Commit:
minor fixes to example 1/?

Commit:
minor fixes to example 2/?

Commit:
example doc string added

Commit:
pep8 improvements

Commit:
aggressive pep8

Commit:
name change

Commit:
travis build fix

Commit:
travis build fix

Commit:
aggressive autopep8

Commit:
aggressive autopep8

Commit:
improve example

Commit:
fix indention for rst

Commit:
line length pep8 fix

Commit:
Fixed multichannel complaints

Commit:
added feature to release doc

Commit:
remove warning from test

jni committed:
Add normalized mutual information metric

jni committed:
Use NMI in the affine registration example

jni committed:
Fix string formatting in affine example

jni committed:
Add journal reference to NMI

jni committed:
Add test for NMI

jni committed:
Add inverse kwarg to register_affine + improve doc

jni committed:
Use NMI as the default metric for registration

jni committed:
Major overhaul of example

jni committed:
Update documentation to use mathjax

jni committed:
Improve compare_nmi functionality and docstring

The function now checks that the input images have the same number of
dimensions, and pads them to be the same size. The documentation now
states the possible range of values for NMI.

jni committed:
Improve NMI docstring

jni committed:
Test and bug fixes for compare_nmi

jni committed:
Add test for random images

jni committed:
Improve registration example

jni committed:
Rename iter_callback to level_callback

jni committed:
Improve the docstring for register_affine

jni committed:
Fix default nlevels

jni committed:
Compute nlevels from simpler arg, minimum image size

jni committed:
Fix maximum pyramid level logic

jni committed:
Fix registration for multichannel images

Add test for registration of multichannel images

jni committed:
Improve example by overlaying transforms

jni committed:
Clarify second parameter to minimize by using keyword

jni committed:
Improve registration example, hand-picking levels

jni committed:
Address reviewer comments

jni committed:
Add check for input image dimensionality

JDWarner committed:
FIX: Astronaut, minor fixes to registration.

jni committed:
Default level_callback is None

jni committed:
Merge branch 'master' into image-registration

jni committed:
Move NMI to new metrics module

jni committed:
Update imports of NMI in registration.py

jni committed:
Add nd support to center and normalize points

jni committed:
Fix broken stability check

jni committed:
Initial work on nD estimation of projective transforms

jni committed:
Untested, but nd projective transform estimation should work

jni committed:
Implement nD affine transforms, sort of

jni committed:
Update gallery example to use metrics

jni committed:
Add support for nD affine and 2-3D similarity

jni committed:
Make projective transforms fully nd

jni committed:
Remove blank line to attempt to fix docstring

jni committed:
Fix docstring and bad valueerror

jni committed:
Use raw string for latex output

jni committed:
dim -> dimensionality; explicit better than implicit

jni committed:
Unify NotImplemented checks in Fundamental and Essential matrix transforms

jni committed:
Update docstring for PiecewiseAffine

jni committed:
Add ability to specify rotation and translation for 3D rigid transforms

jni committed:
Rename 'target' to 'moving' image

jni committed:
Add initial vector to the API

jni committed:
Rename parameters, add translation scale

jni committed:
Allow user to provide vector_to_matrix function

jni committed:
Comment out translation indices for now

jni committed:
Minor formatting fix

jni committed:
Update incorrect dimensionality check in EuclideanTransform

jni committed:
Add __array__ protocol to ProjectiveTransform

jni committed:
Demonstrate how to limit registration to transform classes

jni committed:
Multiply translation parameters by pyramid factor

jni committed:
Refine example with rigid registration

jni committed:
Use partial function instead of closure for cost

jni committed:
Merge branch 'master' into image-registration

jni committed:
Fix incorrect number of arguments test error

jni committed:
Fix another test

jni committed:
Move to new registration module

jni committed:
register_affine -> registration.affine

jni committed:
Minor ReST formatting fix suggested by @JDWarner

jni committed:
Improve affine transform docstring

jni committed:
Add dimensionality to AffineTransform docstring

jni committed:
initial_vector -> initial_parameters

jni committed:
Add comment explaining translation_indices

jni committed:
Minor improvements to docstring

jni committed:
Add newline to fix sphinx build problem

jni committed:
Add reference to Elegant SciPy

jni committed:
Fix normalize and center with degenerate point set

See https://github.com/scikit-image/scikit-image/pull/3544#issuecomment-606391426

jni committed:
Rearrange input arguments

jni committed:
Update default minimum value of minimum pyramid size to 32

jni committed:
Update docstring

jni committed:
Move example to registration

jni committed:
Merge branch 'master' into image-registration

jni committed:
Don't specify stacklevel twice

stacklevel is already specified in our own utils, so adding it is
redundant when using our own warn function.

jni committed:
Update initialisation for translation vector

jni committed:
Fix scaling again

Commit:
Merge branch 'main' into image-registration

Commit:
Update doc/examples/registration/plot_register_affine.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

Commit:
multichannel -> channel_axis

Commit:
Update affine API order according to @stefanv prefs

Commit:
_affine.py: channel_axis -> multichannel

jboulanger committed:
parametric registration with ilk

jboulanger committed:
change parametrization to matrix

jboulanger committed:
Merge remote-tracking branch 'affine/image-registration' into global-registration

jboulanger committed:
add parametric nmi to check that nmi works

jboulanger committed:
change pyramid computation and update api

jboulanger committed:
Unify the handing of the two affine estimation methods using a method string

jboulanger committed:
change string parameter to function

jboulanger committed:
add color support

jboulanger committed:
test weights

jboulanger committed:
parametric -> affine

jboulanger committed:
add doc and add vector_to_matrix to affine

jboulanger committed:
use predefined motion models in affine

Remove translation_indices and _parameter_vector_to_matrix parametrization of the affine function.
Add motion model in lucas kanade affine solver.

Co-authored-by: seanbudd
Co-authored-by: juan.nunez-iglesias@monash.edu

jboulanger committed:
add test_tre

jboulanger committed:
improve docstrings

jboulanger committed:
fix merge goblins

jboulanger committed:
Merge branch 'main' into global-registration

jboulanger committed:
add raises in doc string and fix test_weights

jboulanger committed:
fixing example in affine

jboulanger committed:
fix geometric array like input

jboulanger committed:
fixing example (ci)

jboulanger committed:
replace a@=b by a= a@b

stefanv commented:
Nice example; something to note for maintainers, looks like this exposes new API.

stefanv commented:
@jboulanger This is a huge amount of work; thank you very much! @jni (who guided #3544) and I have been discussing this feature ever since writing Elegant SciPy, and this puts us significantly closer to that goal. I would like for @jni to review. This change is significant enough that we may want to dedicate the next developer call to it, and perhaps you want to join to discuss.

mkcor commented on skimage/metrics/simple_metrics.py:
```diff
@@ -5,16 +5,17 @@
 
 def _as_floats(image0, image1):
     """
     Promote im1, im2 to nearest appropriate floating point precision.
```
```suggestion
    Promote image0, image1 to nearest appropriate floating-point precision.
```

mkcor commented on skimage/metrics/simple_metrics.py:
```diff
@@ -219,6 +221,8 @@ def normalized_mutual_information(image0, image1, *, bins=100):
     bins : int or sequence of int, optional
         The number of bins along each axis of the joint histogram.
+    weights: ndarray | None
+        Weights used in the computation of the histogram
```
```suggestion
        Weights used in the computation of the histogram.
```

mkcor commented on doc/examples/registration/plot_register_affine.py:
```diff
@@ -0,0 +1,187 @@
+
+###############################################################################
+# If we know that our transform is a *rigid* transform, also known as a
+# Euclidean transform we can reduce the number of free parameters in the model.
```
```suggestion
# Euclidean transform, we can reduce the number of free parameters in the model.
```

mkcor commented on skimage/metrics/simple_metrics.py:
```diff
@@ -240,15 +244,15 @@ def normalized_mutual_information(image0, image1, *, bins=100):
-           invariant entropy measure of 3D medical image alignment.
-           Pattern Recognition 32(1):71-86
-           :DOI:`10.1016/S0031-3203(98)00091-0`
+            invariant entropy measure of 3D medical image alignment.
```
When did the syntax/convention change? Just curious. :pray:

jboulanger replied:
Hello, Thank for the modifications. I am not totally sure which syntax/convention you are referring to. Is it in the docstring? I propose a change in simple_metrics which is a bit a side effect of the main proposal. This is to have a uniform support of weights in the registration code across the two implemented methods. I hope this is backward compatible and doesn't break anything.

mkcor replied:
I'm referring to the extra indent, which I also noticed lines 18-19 of `doc/examples/registration/plot_register_affine.py`. Perhaps the formatting got fixed/taken care of by the pre-commit hook, in which case you didn't actually *decide* on it. I was just curious about the change because, up until now(?), we would write:
```rst
.. [1] Juan Nunez-Iglesias, Stefan van der Walt, and Harriet Dashnow. Elegant
       SciPy: The Art of Scientific Python. 1st. O'Reilly Media, Inc.,
       2017. isbn: 1491922877, 9781491922873.
```
and not
```rst
.. [1] Juan Nunez-Iglesias, Stefan van der Walt, and Harriet Dashnow. Elegant
        SciPy: The Art of Scientific Python. 1st. O'Reilly Media, Inc.,
        2017. isbn: 1491922877, 9781491922873.
```

mkcor reviewed (commented):
Just took a quick, superficial look at this huge PR!

jboulanger committed:
Update skimage/metrics/simple_metrics.py

Co-authored-by: Marianne Corvellec <marianne.corvellec@ens-lyon.org>

jboulanger committed:
Update skimage/metrics/simple_metrics.py

Co-authored-by: Marianne Corvellec <marianne.corvellec@ens-lyon.org>

jboulanger committed:
Update doc/examples/registration/plot_register_affine.py

Co-authored-by: Marianne Corvellec <marianne.corvellec@ens-lyon.org>

jboulanger commented:
I have been exchanging a few messages with @jni on tulip before putting this together. However I am aware that the proposition diverged significantly from the original discussion that happened during #3544.  I tried to balance usability and flexibility and also wanted to keep the multi-resolution factorized across the two methods.

jboulanger committed:
update doc and fix reference formating

jboulanger committed:
Merge branch 'global-registration' of https://github.com/jboulanger/scikit-image into global-registration

mkcor commented on skimage/metrics/simple_metrics.py:
```diff
@@ -257,11 +261,18 @@ def normalized_mutual_information(image0, image1, *, bins=100):
 
-    hist, bin_edges = np.histogramdd(
+    if weights is not None and weights.shape != padded0.shape:
+        max_shape = np.maximum(image0.shape, image1.shape)
+        padded_weights = _pad_to(weights, max_shape)
+    else:
+        padded_weights = weights
+
+    hist = np.histogramdd(
         [np.reshape(padded0, -1), np.reshape(padded1, -1)],
         bins=bins,
         density=True,
-    )
+        weights=None if weights is None else np.reshape(padded_weights, -1),
+    )[0]
```
Not sure, but isn't the following more natural to follow? 
```py
    if weights is not None:
        if weights.shape != padded0.shape:
            max_shape = np.maximum(image0.shape, image1.shape)
            padded_weights = _pad_to(weights, max_shape)
        else:
            padded_weights = weights
        weights = np.reshape(padded_weights, -1)

    hist = np.histogramdd(
        [np.reshape(padded0, -1), np.reshape(padded1, -1)],
        bins=bins,
        density=True,
        weights=weights,
    )[0]
```

jboulanger replied:
Yes, it is nicer.

jboulanger committed:
apply suggestion from mkcor

JDWarner commented on skimage/registration/_affine.py:
```diff
@@ -0,0 +1,584 @@
+    options: dict
+        options for scipy.optimization.minimize
+    cost: function
+        Cost function minimize
```
"to be minimized" perhaps?

JDWarner commented on skimage/registration/_affine.py:
```diff
@@ -0,0 +1,584 @@
+    cost: function
+        Cost between registered image and reference with 3 inputs:
+        reference image, moving image and weights.
+    vector_to_matrix:
```
This parameter may be vestigial, it seems this is now automatic.

JDWarner commented on skimage/registration/_affine.py:
```diff
@@ -0,0 +1,584 @@
+        options for scipy.optimization.minimize
+    cost: function
+        Cost function minimize
+    vector_to_matrix: function(param, ndim) -> ndarray
```
This parameter may be vestigial, it seems this is now automatic.

JDWarner commented on skimage/registration/_affine.py:
```diff
@@ -0,0 +1,584 @@
+    matrix : ndarray | None
+        Intial guess for the homogeneous transformation matrix
+    model: str
+        Motion model
```
The docstrings for each solver specify that we can use the model options "affine", "translation" or "euclidean", but most users would interact with this option at the level of this function so we should make that especially clear in this docstring.

As a lower bound, give them the options: `Motion model: "affine", "translation" or "euclidean".` 

In addition, it might be wise to spend a couple sentences expanding on the difference/constraints of a model limited to translation or euclidean registration alone versus affine registration so a user with less experience could choose without searching elsewhere.

jboulanger replied:
I added the following to the affine documentation:
```  
  model: str
        Motion model: "translation", "euclidean" or "affine" where euclidean motion
        corresponds to rigid motion (rotation and translation) and affine motion can
        represent change in scale, shear, rotation and translations.
```

jboulanger replied:
Each function with motion model now has now the doc string
```
Motion model: "affine", "translation" or "euclidean"
```

JDWarner commented on skimage/transform/_geometric.py:
```diff
@@ -58,6 +58,8 @@ def _center_and_normalize_points(points):
 
     """
+    # deal with array_like points
+    points = np.array(points)
```
I think we can use `np.asarray`, which only copies if necessary. In contrast, I believe `np.array` always makes a copy. This could be a performance regression otherwise.

JDWarner reviewed (changes requested):
Thank you for the contribution @jboulanger! What an awesome first PR!

This feature is one I have long anticipated and am excited to see in `scikit-image`.

My review will be in 2 parts; this first set of comments is mostly stylistic from a detailed read-through to get the ball rolling. In the near future, I will follow up after digging in further with testing on real images in both 2 and 3 dimensions, as well as some performance profiling.

Thanks again!

jboulanger commented on skimage/registration/_affine.py:
```diff
@@ -0,0 +1,584 @@
+
+    """
+    # Create a regular set of points on the grid
+    points = np.concatenate(
+        [
+            np.stack(
+                [
+                    x.flatten()
+                    for x in np.meshgrid(*[np.arange(n) for n in shape], indexing="ij")
+                ],
+                axis=1,
+            ).T,
+            np.array([1] * np.prod(shape)).reshape(1, -1),
+        ]
+    )
```
This part was not great indeed
```
slc = [slice(0,n) for n in shape]
N = np.prod(shape)
points = np.stack([*[x.ravel() for x in np.mgrid[slc]],np.ones(N)])
```

jboulanger commented on skimage/registration/_affine.py:
```diff
@@ -0,0 +1,584 @@
+        T = np.eye(ndim + 1)
+        T[:ndim, -1] = parameter[:ndim].ravel()
+        matrix = matrix @ T
+        T = np.array([[1, 1, parameter[0]], [0, 1, parameter[1]], [0, 0, 1]])
```
That's correct.

jboulanger commented on skimage/registration/_affine.py:
```diff
@@ -0,0 +1,584 @@
+        # Rotations for each planes
+        for k, a in enumerate(combinations(range(ndim), 2)):
+            c, s = cos(parameter[ndim + k]), sin(parameter[ndim + k])
+            R = np.eye(ndim + 1)
+            R[a[0], a[0]] = c
+            R[a[0], a[1]] = -s
+            R[a[1], a[1]] = c
+            R[a[1], a[0]] = s
+            matrix = matrix @ R
+        # Translation along each axis
+        T = np.eye(ndim + 1)
+        T[:ndim, -1] = parameter[:ndim].ravel()
+        matrix = matrix @ T
```
I renamed them to rot and trans, as they represent rotation matrices and translation vectors.

jboulanger commented on skimage/registration/_affine.py:
```diff
@@ -0,0 +1,584 @@
+    for i, j in combinations_with_replacement(range(len(elems)), 2):
+        G[i, j] = G[j, i] = (elems[i] * elems[j] * weights).sum()
+
+    # Model reduction
+    if model.lower() == "translation":
+        H = np.eye(G.shape[0], ndim, dtype=np.float64)
+    elif model.lower() == "euclidean":
+        if ndim == 2:
+            H = np.array(
+                [
+                    [1.0, 0.0, 0.0],
+                    [0.0, 1.0, 0.0],
+                    [0.0, 0.0, 0.0],
+                    [0.0, 0.0, -1.0],
+                    [0.0, 0.0, 1.0],
+                    [0.0, 0.0, 0.0],
+                ],
+            )
+        else:
+            raise NotImplementedError("Eulidean motion model implemented only in 2D")
+    elif model.lower() == "affine":
+        H = np.eye(ndim * (ndim + 1), dtype=matrix.dtype)
+    else:
+        raise NotImplementedError(f"Unsupported motion model {model}")
+
+    for _ in range(max_iter):
+        # Warp each channel
+        moving_image_warp = np.stack(
+            [ndi.affine_transform(plane, matrix) for plane in moving_image]
+        )
+
+        # Compute the error (eq. (26))
+        error_image = reference_image - moving_image_warp
+
+        # Compute b from eq. (33)
+        b = np.array([[(e * error_image * weights).sum()] for e in elems])
+
+        try:
+            # Solve the system
+            r = np.linalg.solve(H.T @ G @ H, H.T @ b)
```
I renamed the two variables as proposed.

jboulanger commented on skimage/registration/_affine.py:
```diff
@@ -0,0 +1,584 @@
+            if np.linalg.norm(r) < tol:
+                break
+
+        except np.linalg.LinAlgError:
```
I propose this warning:
```
    warnings.warn("Failed to invert linear system (np.linalg.solve).")
```

jboulanger commented on skimage/registration/_affine.py:
```diff
@@ -0,0 +1,584 @@
+    moving_image,
+    weights,
+    *,
+    cost=cost_nmi,
```
This part was developed originally in another PR ( @jni  @seanbudd ). I integrated the suggestion to use a lambda, removed cost_nmi and redefined the default parameters in _param_cost.

jboulanger commented on skimage/registration/_affine.py:
```diff
@@ -0,0 +1,584 @@
+        model=model,
+    )
+
+    for J0, J1, W in pyramid[1:]:
```
In order to be more explicit:
```
 for image_0, image_1, w in pyramid[1:]:
```

jboulanger commented on skimage/registration/tests/test_affine.py:
```diff
@@ -0,0 +1,156 @@
+def create_matrix(shape, model, ndim):
+    """Create a homogeneous test matrix"""
+    # Center the transformations
+    T = np.eye(ndim + 1, dtype=np.float64)
+    T[:ndim, -1] = np.array(shape) / 2
+
+    matrix = np.eye(ndim + 1, dtype=np.float64)
+    if model == "translation":
+        matrix[:ndim, -1] += np.random.uniform(-10, 10, size=(ndim))
+    elif model == "euclidean":
+        R = np.eye(ndim + 1, dtype=np.float64)
+        # Rotations for each planes
+        for k, a in enumerate(combinations(range(ndim), 2)):
+            r = np.random.uniform(-0.1, 0.1)
+            c, s = math.cos(r), math.sin(r)
+            R[a[0], a[0]] = c
+            R[a[0], a[1]] = -s
+            R[a[1], a[1]] = c
+            R[a[1], a[0]] = s
+            matrix = matrix @ R
+        matrix[:ndim, -1] += np.random.uniform(-1, 1, size=(ndim))
+    else:
+        matrix[:ndim, :] += np.random.uniform(-0.01, 0.01, size=(ndim, ndim + 1))
+
+    return np.linalg.inv(T) @ matrix @ T
```
Renamed as rot and trans.

jboulanger commented on skimage/registration/__init__.pyi:
```diff
@@ -2,7 +2,23 @@
+from ._affine import (
+    affine,
+    studholme_affine_solver,
+    lucas_kanade_affine_solver,
```
I'll wait a bit for more feedback on this. I have no big preference.

jboulanger replied:
Hello, I waited a long time before finally renaming the solvers solver_affine_lucas_kanade and solver_affine_studholme.

lagru replied:
Sorry for keeping you waiting @jboulanger. :pray: Maintainers are sometimes a bit short on time and then items like this quickly leave the radar. Large PRs like this also require a bit of time to review which makes things more difficult if all you have is a few minutes on the side. But that doesn't mean we don't want and eventually will get contributions merged. :crossed_fingers: 

I haven't had time to look into this PR in more depth but still wanted to clarify. :)

jboulanger commented on skimage/registration/_affine.py:
```diff
@@ -0,0 +1,584 @@
+    *,
+    method="Powell",
+    options={"maxiter": 10, "disp": False},
+    cost=cost_nmi,
```
This part was developed originally in another PR ( @jni @seanbudd ). I integrated the suggestion to use a lambda, removed cost_nmi and redefined the default parameters in _param_cost.

jboulanger commented on skimage/registration/_affine.py:
```diff
@@ -0,0 +1,584 @@
+
+def target_registration_error(shape, matrix):
+    """
+    Compute the displacement norm of the transform at at each pixel
```
I hope I've been through all of the incomplete sentences.

jboulanger committed:
implement suggestions from JDWarner (documentation and style)

jboulanger commented:
Thank you for the very informative feedback!

jboulanger committed:
Merge branch 'scikit-image:main' into global-registration

jboulanger committed:
correct parameter scaling from initial matrix parameter. Add corresponding test.

jboulanger committed:
Merge branch 'scikit-image:main' into global-registration

jboulanger committed:
rename solvers following @JDWarner comment

jboulanger committed:
remove cost_nmi that is not used anymore

jboulanger committed:
Merge branch 'main' into global-registration

jboulanger committed:
Merge branch 'main' into global-registration

jboulanger committed:
Merge branch 'main' into global-registration

mkcor committed:
Use src layout and create skimage2 directory. (#7864)

## Description

As per our discussion at the in-person sprint, let's go for the `src`
layout and get started with the `skimage2` namespace.

/cc @lagru @stefanv @jni @matthew-brett @Czaki 

## Checklist

<!-- Before pull requests can be merged, they should provide: -->

- A descriptive but concise pull request title
- [Docstrings for all
functions](https://github.com/numpy/numpydoc/blob/main/doc/example.py)
- [Unit
tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution
guide](https://scikit-image.org/docs/dev/development/contribute.html) is
followed

## Release note

For maintainers and optionally contributors, please refer to the
[instructions](https://scikit-image.org/docs/stable/development/contribute.html#documenting-changes)
on how to document this PR for the release notes.

```release-note
...
```

---------

Co-authored-by: Stefan van der Walt <stefanv@berkeley.edu>
Co-authored-by: Lars Grüter <lagru@mailbox.org>
Co-authored-by: Grzegorz Bokota <bokota@gmail.com>
Co-authored-by: Stefan van der Walt <stefan@mentat.za.net>
Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

mkcor committed:
Merge conflicts

mkcor committed:
Merge branch 'main' of github.com:scikit-image/scikit-image

mkcor committed:
Merge branch 'main' of github.com:scikit-image/scikit-image

mkcor committed:
Merge branch 'main' of github.com:scikit-image/scikit-image

mkcor committed:
Add affine function for image registration

Co-authored-by: Jérôme Boulanger <jemore.boulanger@gmail.com>
Co-authored-by: Juan Nunez-Iglesias <jni@fastmail.com>
Co-authored-by: Sean Budd <seanbudd@me.com>
Co-authored-by: Sean Budd <sean@nvaccess.org>

mkcor committed:
Merge branch 'main' into resubmit-pr-7421

jboulanger committed:
Reformulate the introduction text in the affine registration example.

loicmusy committed:
First rough implementation of the ECC Solver into the affine method implemented by @jboulanger

jboulanger committed:
update plot_register_affine

jboulanger committed:
Merge remote-tracking branch 'ecc/ecc' into global-registration

jboulanger committed:
improve ecc documentation and fix staility of studholm

jboulanger committed:
fix scaling issues in studholme

jboulanger committed:
fix test replaing equality test with almost equal

jboulanger committed:
change scale_parameters to scale_matrix and fix test and examples

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
